src/plots/grid.js

```diff
@@ -0,0 +1,409 @@
+            'If you give an `xaxes` array, its length is used as the default.',
+            'But it\'s also possible to have a different length, if you',
+            'want to leave a row at the end for non-cartesian subplots.'
+        ].join(' ')
+    },
+    subplots: {
+        valType: 'info_array',
+        freeLength: true,
+        dimensions: 2,
+        items: {valType: 'enumerated', values: [counterRegex('xy').toString(), '']},
+        role: 'info',
+        editType: 'calc',
+        description: [
+            'Used for freeform grids, where some axes may be shared across subplots',
+            'but others are not. Each entry should be a cartesian subplot id, like',
+            '*xy* or *x3y2*, or ** to leave that cell empty. You may reuse x axes',
+            'within the same column, and y axes within the same row.',
+            'Non-cartesian subplots and traces that support `domain` can place themselves',
+            'in this grid separately using the `gridcell` attribute.'
+        ].join(' ')
+    },
+    xaxes: {
+        valType: 'info_array',
+        freeLength: true,
+        items: {valType: 'enumerated', values: [cartesianIdRegex.x.toString(), '']},
+        role: 'info',
+        editType: 'calc',
+        description: [
+            'Used with `yaxes` when the x and y axes are shared across columns and rows.',
+            'Each entry should be an x axis id like *x*, *x2*, etc., or ** to',
+            'not put an x axis in that column. Entries other than ** must be unique.',
+            'Ignored if `subplots` is present. If missing but `yaxes` is present,',
+            'will generate consecutive IDs.'
+        ].join(' ')
+    },
+    yaxes: {
+        valType: 'info_array',
+        freeLength: true,
+        items: {valType: 'enumerated', values: [cartesianIdRegex.y.toString(), '']},
+        role: 'info',
+        editType: 'calc',
+        description: [
+            'Used with `yaxes` when the x and y axes are shared across columns and rows.',
+            'Each entry should be an y axis id like *y*, *y2*, etc., or ** to',
+            'not put a y axis in that row. Entries other than ** must be unique.',
+            'Ignored if `subplots` is present. If missing but `xaxes` is present,',
+            'will generate consecutive IDs.'
+        ].join(' ')
+    },
+    pattern: {
+        valType: 'enumerated',
+        values: ['independent', 'coupled'],
+        dflt: 'coupled',
+        role: 'info',
+        editType: 'calc',
+        description: [
+            'If no `subplots`, `xaxes`, or `yaxes` are given but we do have `rows` and `columns`,',
+            'we can generate defaults using consecutive axis IDs, in two ways:',
+            '*coupled* gives one x axis per column and one y axis per row.',
+            '*independent* uses a new xy pair for each cell, left-to-right across each row',
+            'then iterating rows according to `roworder`.'
+        ].join(' ')
+    },
+    gap: {
```

Comment:
I used `xgap` and `ygap` to match `heatmap`, but I felt like it's also useful to have a symmetric `gap` attribute (that's only used to set the defaults for `xgap` and `ygap`) - thoughts? Keep this (and add it to `heatmap`), or remove it and just use `xgap`/`ygap`?

Also, for completeness, in other contexts I've seen either `grout` or `gutter` used for this effect, do we want to stick with `gap` or use one of these?

Reply:
Sounds good, I'll 🔪 `gap` and keep `xgap` and `ygap`.

The meaning of horizontal/vertical spacing in the python wrapper is different from what I have for `xgap`/`ygap` - the python version seems to use absolute domain fraction (hence the need to divide the default by the number of cols/rows, whereas in my current implementation it's a fraction of the grid period, ie `xgap = xgapPx / (xgapPx + xplotPx)`. That allows me to use a constant value for the default, and I feel like it's easier to use as you can think of how the gap compares to one cell of the grid, rather than how it compares to the whole plot area... do you agree or should I switch to what we did in python?

Reply:
I prefer your implementation :ok_hand:

Reply:
🔪 gap and made the defaults (for independent axes, not for coupled) match the python version 0.2 and 0.3 instead of the 0.25 for both I had previously. -> https://github.com/plotly/plotly.js/pull/2399/commits/670bdd59bb696fb5ef6bf0d785a183632451ba8c